File: Level2/스킬트리.cpp
#include <string>
#include <vector>
#include <stack>
#include <algorithm>
#include <queue>
using namespace std;

int solution(string skill, vector<string> skill_trees)
{
	int answer = 0;
	int check = 0;
	queue<int> test;
	queue<int> test2;
	for (int t = 0; t < skill.length(); t++)
		test2.push(skill[t]);

	for (int j = 0; j < skill_trees.size(); j++)
	{
		for (int i = 0; i < skill_trees[j].length(); i++)
		{
			for (int p = 0; p < skill.length(); p++)
			{
				if (skill_trees[j][i] == skill[p])
					test.push(skill[p]);
			}
		}
		if (test.size() == 0)
			answer++;
		else if (test.front() != test2.front())
		{
			while (!test.empty())
				test.pop();
		}
		else if(test.front() == test2.front())
		{
			test.pop();
			test2.pop();
			while (!test.empty() && !test2.empty())
			{
				if (test.front() != test2.front())
				{
					break;
				}
				else if (test.front() == test2.front())
				{
					test.pop();
					test2.pop();
				}
			}
			if (test.empty() && test2.empty())
			{
				answer++;
				for (int i = 0; i < skill.length(); i++)
					test2.push(skill[i]);
			}
			else if (!test.empty() && test2.empty())
			{
					answer++;
					while (!test.empty())
						test.pop();
					for (int t = 0; t < skill.length(); t++)
						test2.push(skill[t]);
			}
			else if (test.empty() && !test2.empty())
			{
				answer++;
				while (!test2.empty())
					test2.pop();
				for (int t = 0; t < skill.length(); t++)
					test2.push(skill[t]);
			}
			else
			{
				while (!test2.empty())
					test2.pop();
				for (int t = 0; t < skill.length(); t++)
					test2.push(skill[t]);
				while (!test.empty())
					test.pop();
			}
		}
	}
	return answer;
}

File: Level2/스킬트리.py
skill = "CBD"
skill_trees = ["BACDE", "CBADF", "AECB", "BDA"]


def solution(skill, skill_trees):
    lis = []
    answer = ''
    anl = []
    correct = [0] * len(skill_trees)
    count = 0
    for i in range(len(skill)):
        lis.append(skill[i])
    for k in skill_trees:
        for i in range(len(k)):
            if k[i] in lis:
                answer += k[i]
        anl.append(answer)
        answer = ''

    for i in range(len(skill)):
        tmp = skill[:i+1]
        for j in range(len(anl)):
            if tmp == anl[j] and correct[j] != -1:
                count += 1
                correct[j] = -1
    for j in anl:
        pri = 0
        for k in range(len(j)):
            if j[k] in skill:
                pri += 1
        if pri == 0:
            count += 1
    return count



did = solution(skill, skill_trees)
print(did)
